libs/langchain/src/agents/middlewareAgent/middleware/anthropicCodeExecution.ts

```diff
@@ -0,0 +1,215 @@
+});
+
+export type AnthropicCodeExecutionMiddlewareState = z.infer<typeof stateSchema>;
+
+/**
+ * Creates middleware for managing Anthropic's code execution tool.
+ *
+ * This middleware streamlines data analysis workflows by:
+ * - Enabling Anthropic's code execution tool
+ * - Automatically managing container IDs for Anthropic (enables file persistence across turns)
+ * - Injecting required beta headers for code execution and files API
+ * - Working around Anthropic API bugs with file references in conversation history
+ *
+ * Works with:
+ * - Anthropic's `code_execution_20250825` tool with Files API
+ *
+ * @example
+ * ```typescript
+ * import { ChatAnthropic } from "@langchain/anthropic";
+ * import { HumanMessage } from "@langchain/core/messages";
+ * import { createAgent } from "langchain/agents";
+ * import { anthropicCodeExecutionMiddleware } from "langchain/agents/middleware";
+ * import {
+ *   uploadFileAnthropic,
+ *   extractGeneratedFilesAnthropic,
+ *   getFileMetadataAnthropic,
+ *   downloadFileAnthropic,
+ * } from "langchain/agents/middleware";
+ *
+ * const model = new ChatAnthropic({
+ *   model: "claude-sonnet-4-20250514",
+ * });
+ *
+ * const agent = createAgent({
+ *   model,
+ *   middleware: [anthropicCodeExecutionMiddleware()],
+ * });
+ *
+ * // Upload file using helper
+ * const client = model.createClient({});
+ * const uploadedFile = await uploadFileAnthropic(client, "data.csv");
+ *
+ * // Analyze data
+ * const result = await agent.invoke({
+ *   messages: new HumanMessage({
+ *     content: [
+ *       { type: "text", text: "Analyze this data" },
+ *       { type: "container_upload", file_id: uploadedFile.fileId },
+ *     ],
+ *   }),
+ * });
+ *
+ * // Extract and download generated files
+ * const fileIds = extractGeneratedFilesAnthropic(result);
+ * for (const fileId of fileIds) {
+ *   const metadata = await getFileMetadataAnthropic(client, fileId);
+ *   await downloadFileAnthropic(client, fileId, metadata.filename);
+ * }
+ * ```
+ *
+ * @returns A configured middleware instance
+ *
+ * @see {@link AnthropicCodeExecutionMiddlewareState} for the state schema
+ * @see {@link uploadFileAnthropic} for uploading files to Anthropic
+ * @see {@link extractGeneratedFilesAnthropic} for extracting generated file IDs from responses
+ * @see {@link getFileMetadataAnthropic} for retrieving file metadata including filenames
+ * @see {@link downloadFileAnthropic} for downloading generated files from Anthropic
+ */
+export function anthropicCodeExecutionMiddleware() {
+  return createMiddleware({
+    name: "anthropicCodeExecutionMiddleware",
+    stateSchema,
+    tools: [{ type: "code_execution_20250825", name: "code_execution" }],
+    modifyModelRequest: (request, state) => ({
+      ...request,
+      messages: workAroundAnthropicCodeExecutionBug(request.messages),
+      callOptions: {
+        // Pass container ID to reuse files across turns
+        // FIXME: What to do about container expiration?
```

Comment:
wasn't sure how to handle container expiration. Open to suggestions, including punt